README.md

Ranked Choice Voting — R Package
================================

[![Build Status](https://travis-ci.org/ds-elections/rcv.svg?branch=master)](https://travis-ci.org/ds-elections/rcv)

`rcv` helps you work directly with raw ballot image and cast vote record data to run elections.

**Features**

-   Read in ballot image and master lookup files
-   Merge these files to get a "readable" ballot layout
-   Conduct elections, and view a round-by-round table of results
-   Visualize the flow of voters with an interactive Sankey diagram
-   Compatible with `dplyr`/`magrittr` pipe syntax (`%>%`)

**Installation**

We're on [CRAN](https://cran.r-project.org/package=rcv) now! You can install our first release (0.2.0):

``` r
install.packages("rcv")
```

Or, you can install the development version of `rcv` here:

``` r
devtools::install_github("ds-elections/rcv")
```

Created by:

-   Jay Lee, <jaylee@reed.edu>
-   Matthew Yancheff, <yanchefm@reed.edu>

The style of this README is inspired by the [googlesheets](%22github.com/jennybc/googlesheets%22) R package.

Basic Workflow Demo
-------------------

`sf_bos_ballot` is included as an example raw ballot image, and `sf_bos_lookup` as an example raw master lookup. Both are included as `.rda`s, and they are in the "WinEDS" format. This data comes from the 2016 San Francisco Board of Supervisors elections ([San Francisco Department of Elections](http://www.sfelections.org/results/20161108/#english_detail)).

``` r
head(sf_bos_ballot)
```

    ## # A tibble: 6 x 1
    ##                                              X1
    ##                                           <chr>
    ## 1 000000900000660300000010020000406001000012800
    ## 2 000000900000660300000010020000406002000000001
    ## 3 000000900000660300000010020000406003000000001
    ## 4 000000900000660400000010020000406001000012300
    ## 5 000000900000660400000010020000406002000012500
    ## 6 000000900000660400000010020000406003000012100

``` r
head(sf_bos_lookup)
```

    ## # A tibble: 6 x 1
    ##                                                                            X1
    ##                                                                         <chr>
    ## 1 Candidate 0000121SAMUEL KWONG                                      00000010
    ## 2 Candidate 0000131TIM E. DONNELLY                                   00000010
    ## 3 Candidate 0000133DEAN PRESTON                                      00000010
    ## 4 Candidate 0000135JOEL ENGARDIO                                     00000010
    ## 5 Candidate 0000140MELISSA SAN MIGUEL                                00000010
    ## 6 Candidate 0000144AHSHA SAFAI                                       00000010

#### Cleaning Data

The streamlined version of this process is done with the `clean_ballot()` function. `b_header` and `l_header` are logical values, based on whether the ballot and lookup file respectively have a header for the first row.

``` r
cleaned <- clean_ballot(ballot = sf_bos_ballot, b_header = T, 
                        lookup = sf_bos_lookup, l_header = T, 
                        format = "WinEDS")
knitr::kable(head(readable(cleaned)))
```

| contest                          | pref\_voter\_id | 1                | 2                | 3                |
|:---------------------------------|:----------------|:-----------------|:-----------------|:-----------------|
| Board of Supervisors, District 1 | 000006603       | SANDRA LEE FEWER | NA               | NA               |
| Board of Supervisors, District 1 | 000006604       | MARJAN PHILHOUR  | DAVID LEE        | SAMUEL KWONG     |
| Board of Supervisors, District 1 | 000006605       | DAVID LEE        | RICHIE GREENBERG | BRIAN J. LARKIN  |
| Board of Supervisors, District 1 | 000006606       | MARJAN PHILHOUR  | DAVID LEE        | SANDRA LEE FEWER |
| Board of Supervisors, District 1 | 000006607       | BRIAN J. LARKIN  | ANDY THORNLEY    | JASON JUNGREIS   |
| Board of Supervisors, District 1 | 000006608       | MARJAN PHILHOUR  | NA               | NA               |

To access intermediate steps, the following process can be used.

``` r
# Import and label ballot image
a <- import_data(data = sf_bos_ballot, header = T) %>%
    label(image = "ballot", format = "WinEDS")

# Import and label master lookup
b <- import_data(data = sf_bos_lookup, header = T) %>%
    label(image = "lookup", format = "WinEDS")

# Merge these two tables
c <- characterize(ballot = a, lookup = b, format = "WinEDS")

knitr::kable(head(readable(c)))
```

| contest                          | pref\_voter\_id | 1                | 2                | 3                |
|:---------------------------------|:----------------|:-----------------|:-----------------|:-----------------|
| Board of Supervisors, District 1 | 000006603       | SANDRA LEE FEWER | NA               | NA               |
| Board of Supervisors, District 1 | 000006604       | MARJAN PHILHOUR  | DAVID LEE        | SAMUEL KWONG     |
| Board of Supervisors, District 1 | 000006605       | DAVID LEE        | RICHIE GREENBERG | BRIAN J. LARKIN  |
| Board of Supervisors, District 1 | 000006606       | MARJAN PHILHOUR  | DAVID LEE        | SANDRA LEE FEWER |
| Board of Supervisors, District 1 | 000006607       | BRIAN J. LARKIN  | ANDY THORNLEY    | JASON JUNGREIS   |
| Board of Supervisors, District 1 | 000006608       | MARJAN PHILHOUR  | NA               | NA               |

The `readable()` function takes the clean image, which is formatted for ease in computation, and formats it to be easily read manually.

#### Running Elections

This is done with the `rcv_tally()` function. `sf_bos_clean` is included as an example of a pre-cleaned ballot using the functions above. We will run the District 1 election from this ballot image.

``` r
results <- rcv_tally(sf_bos_clean, "Board of Supervisors, District 1")
knitr::kable(results)
```

| candidate          |  round1|  round2|  round3|  round4|  round5|  round6|  round7|  round8|  round9|
|:-------------------|-------:|-------:|-------:|-------:|-------:|-------:|-------:|-------:|-------:|
| SANDRA LEE FEWER   |   12550|   12689|   12777|   12840|   13029|   13093|   13225|   13354|   14705|
| MARJAN PHILHOUR    |   11067|   11135|   11247|   11348|   11487|   11680|   11837|   12086|   13126|
| DAVID LEE          |    3396|    3408|    3488|    3551|    3622|    3857|    3961|    4093|      NA|
| RICHIE GREENBERG   |     974|     984|    1042|    1220|    1272|    1386|    1508|      NA|      NA|
| BRIAN J. LARKIN    |     747|     773|     832|     896|     956|     997|      NA|      NA|      NA|
| SAMUEL KWONG       |     740|     744|     760|     785|     814|      NA|      NA|      NA|      NA|
| JONATHAN LYENS     |     609|     652|     679|     726|      NA|      NA|      NA|      NA|      NA|
| JASON JUNGREIS     |     611|     626|     654|      NA|      NA|      NA|      NA|      NA|      NA|
| SHERMAN R. D'SILVA |     557|     566|      NA|      NA|      NA|      NA|      NA|      NA|      NA|
| ANDY THORNLEY      |     359|      NA|      NA|      NA|      NA|      NA|      NA|      NA|      NA|
| NA                 |    3499|    3532|    3630|    3743|    3929|    4096|    4578|    5576|    7278|

Sandra Lee Fewer wins in Round 9, with 14,705 votes to Marjan Philhour's 13,126. 3,499 ballots were left blank for this election, and 3,779 were exhausted or marked invalid in this election.

#### Visualizing Data

We have two recommended methods of visualizing RCV data. Both utilize a flowchart called a "Sankey diagram" to show the transfer of voters between rounds. We will use each method to visualize the transfer of voters in the San Francisco District 7 Board of Supervisors election, because District 1 has too many crossings to be readable.

Method 1 (preferred because it is interactive, quicker, and more readable) uses the `networkD3` package:

``` r
d3_7 <- rcv::make_d3list(results = sf_7_results)
networkD3::sankeyNetwork(Links = d3_7$values, Nodes = d3_7$names,
                         Source = "source", Target = "target",
                         Value = "value", NodeID = "candidate", units = "voters",
                         fontSize = 12, nodeWidth = 20)
```

(Picture removed for package update)

Method 2 uses the `alluvial` package (this type of graphic is also called an alluvial diagram):

``` r
# eval set to FALSE for CRAN release

alluvial_7 <- rcv::make_alluvialdf(image = sf_bos_clean,
                                   rcvcontest = "Board of Supervisors, District 7")
alluvial::alluvial(
  alluvial_7[,1:4], freq = alluvial_7$frequency,
  col = ifelse(alluvial_7$round4 == "NORMAN YEE", "lightgreen", "gray"),
  border = "gray", alpha = 0.7, blocks = TRUE
)
```


NEWS.md
# rcv 0.2.1

* Fixed to work with tidyr v0.7.0


R/tabulation.R
#' Determines RCV round results in a dataframe
#'
#' @param image A dataframe containing rcv election data
#' @param rcvcontest (optional) The election to calculate results for. If the image
#' contains more than one unique contest, this must be supplied. In most cases
#' except for some San Francisco elections, this is unnecessary.
#' @return A dataframe that contains vote tallies
#' @examples
#' rcv_tally(image = sf_bos_clean, rcvcontest = "Board of Supervisors, District 7")
#' @export
rcv_tally <- function(image, rcvcontest) {
  contest <- candidate <- pref_voter_id <- vote_rank <- n <- total <- NULL
  unique.ballot.candidate. <- . <- NULL
  if (!(missing(rcvcontest))) {
    image <- image %>% dplyr::filter(contest == rcvcontest)
  }
  ballot <- image %>%
    dplyr::filter(stringr::str_detect(candidate, "=") %in% c(F, NA)) %>%
    dplyr::mutate(candidate = ifelse(is.na(candidate),
                                     "NA",
                                     candidate)) %>%
    dplyr::select(pref_voter_id,
                  vote_rank,
                  candidate)

  n.cand <- length(unique(ballot$candidate))
  results <- data.frame(matrix(rep(NA, n.cand*(n.cand-2)),
                               nrow = n.cand))
  roundnames <- c()
  for (i in 1:(n.cand - 2)) {
    roundnames <- append(roundnames, paste0("round", i))
  }
  colnames(results) <- roundnames
  row.names(results) <- unique(ballot$candidate)

  elim <- data.frame(candidate = character())

  for (j in 1:(n.cand - 2)) {
    if (j >= 2) {
      transfers <- ballot %>%
        dplyr::filter(vote_rank == min(vote_rank),
                      candidate %in% loser) %>%
        dplyr::select(pref_voter_id)
      a <- 0
    } else transfers <- ballot %>% dplyr::select(pref_voter_id)

    if (nrow(transfers) == 0) {
      transfers <- ballot %>% dplyr::select(pref_voter_id)
      a <- 1
    }

    ballot <- ballot %>%
      dplyr::filter(!(candidate %in% elim$candidate))

    candidates <- data.frame(unique(ballot$candidate)) %>%
      dplyr::transmute(candidate = as.character(unique.ballot.candidate.))

    round <- ballot %>%
      dplyr::filter(pref_voter_id %in% transfers$pref_voter_id) %>%
      dplyr::group_by(pref_voter_id) %>%
      dplyr::filter(vote_rank == min(vote_rank)) %>%
      dplyr::ungroup() %>%
      dplyr::group_by(candidate) %>%
      dplyr::summarise(total = n()) %>%
      data.frame() %>%
      dplyr::right_join(candidates,
                 by = c("candidate")) %>%
        dplyr::mutate(total = ifelse(is.na(total), 0, total))

    row.names(round) <- round$candidate

    for (i in unique(round$candidate)) {
      if (j >= 2) {
        if (a != 1) {
          round[i, 2] <- (results[i, j-1] + round[i, 2])
        }
      }

      results [i, j] <- round[i, 2]

    }

    loser <- round %>%
      dplyr::filter(candidate != "NA") %>%
      dplyr::filter(total == min(total)) %>%
      dplyr::select(candidate)

    elim <- rbind(elim, loser)
  }

  results %>%
    dplyr::select(which(as.vector(colSums(is.na(.)) < (nrow(.) - 1)))) %>%
    tibble::rownames_to_column("candidate") %>%
    dplyr::arrange(candidate == "NA",
                   rowSums(is.na(.)),
                   dplyr::desc(.[ ,ncol(.)])) %>%
    tibble::column_to_rownames("candidate") %>%
    add_exhausted()

}


#' Adds correct exhausted numbers for a `rcv_tally` dataframe
#'
#' @param results A dataframe that comes from `rcv_tally()`
#' @return The results dataframe with correct counts for the exhausted votes
#' @export
add_exhausted <- function(results) {

  total <- sum(results[, 1], na.rm = T)
  exhausted <- data.frame(matrix(rep(NA, ncol(results)), nrow = 1))
  colnames(exhausted) <- colnames(results)
  row.names(exhausted) <- c("Exhausted")
  for (i in 1:ncol(results)) {
    exhausted[1, i] <- total - sum(results[, i], na.rm = T)
    exhausted[1, i] <- sum(exhausted[1, i], results["NA", i], na.rm = T)
  }
  results["NA", ] <- exhausted[1, ]
  results <- results %>%
    tibble::rownames_to_column("candidate")

  return(results)
}


R/document_data.R
#' Raw ballot image data from a San Francisco RCV election
#'
#' The .rda version of a raw .txt file, containing ballot data
#' in numeric form for the 2016 San Francisco Board of Supervisors
#' elections in Districts 1, 3, 5, 7, 9, and 11.
#'
#' @format A data frame with 643806 rows and 1 variable
#' @source \url{http://www.sfelections.org/results/20161108/data/20161206/20161206_ballotimage.txt}
"sf_bos_ballot"

#' Raw ballot lookup data from a San Francisco RCV election.
#'
#' The .rda version of a raw .txt file, containing numeric and
#' character information to match the numeric data in the raw
#' ballot image with the objects these numeric strings refer to.
#' Data refers to the 2016 San Francisco Board of Supervisors
#' elections in Districts 1, 3, 5, 7, 9, and 11. All San Francisco
#' data is to be used with the "WinEDS" format.
#'
#' @format A data frame with 644 rows and 1 variable
#' @source \url{http://www.sfelections.org/results/20161108/data/20161206/20161206_masterlookup.txt}
"sf_bos_lookup"

#' Cleaned ballot image data from a San Francisco RCV election
#'
#' A tidied version of `sf_bos_ballot` in a "tall" format,
#' readable to see voters' information, including candidate
#' rankings, precinct, and over-/under-vote codings. Cleaned
#' with clean_ballot.
#' clean_ballot(sf_bos_ballot, b_header = T,
#'              sf_bos_lookup, l_header = T,
#'              format = "WinEDS")
#'
#' @format A data frame with 643806 rows and 9 variables:
#' \describe{
#'   \item{contest}{which election this candidate ranking applies to}
#'   \item{pref_voter_id}{a unique key identifying an individual voter }
#'   \item{serial_number}{the serial number of the voting machine used}
#'   \item{tally}{contains information about whether the ballot was cast
#'   in early voting, was filed provisionally, and other factors}
#'   \item{precinct}{which city precinct the voter voted in}
#'   \item{vote_rank}{the rank given to the candidate by the voter; in
#'   the San Francisco case can take values of 1, 2, or 3}
#'   \item{candidate}{the chosen candidate for the specified vote rank}
#'   \item{over_vote}{a dummy variable, coded 1 if the ballot shows more
#'   votes cast than the number of candidates to be elected}
#'   \item{under_vote}{a dummy variable, coded 1 if the ballot shows no
#'   valid selection made for a candidate}
#'   }
"sf_bos_clean"

#' Raw ballot image data from a Cambridge, MA RCV election
#'
#' The .rda version of a .csv from a 2005 Cambridge, MA City Council
#' multiwinner RCV election. This data is included as an example of
#' another ballot image format for use with the label function. All
#' Cambridge data is to be used with the "ChoicePlus" format.
#'
#' @format A data frame with 17959 rows and 26 variables
"cambridge_ballot"

#' Raw lookup data from a Cambridge, MA RCV election
#'
#' The .rda version of a .csv from a 2005 Cambridge, MA City Council
#' multiwinner RCV election. This data is included as an example of another
#' master lookup format for use with the label function. This data includes
#' information other than the candidate names and codes, but we only use the
#' candidate names and codes.
#'
#' @format A data frame with 588 rows and 6 variables
"cambridge_lookup"

#' Cleaned ballot image data from a San Francisco RCV election
#'
#' A tidied version of `cambridge_ballot` in a "tall" format,
#' readable to see voters' information, including candidate
#' rankings, ward, and precinct. Cleaned with clean_ballot.
#' clean_ballot(cambridge_ballot, b_header = T,
#'              cambridge_lookup, l_header = T,
#'              format = "ChoicePlus")
#'
#' @format A data frame with 108752 rows and 7 variables:
#' \describe{
#'   \item{pref_voter_id}{a unique key identifying an individual voter }
#'   \item{ward}{which city ward the voter voted in}
#'   \item{precinct}{which city precinct the voter voted in}
#'   \item{style}{which ballot style the voter had; different styles list
#'   different candidates first on the ballot to remove that advantage}
#'   \item{contest}{which election this candidate ranking applies to}
#'   \item{vote_rank}{the rank given to the candidate by the voter}
#'   \item{candidate}{the chosen candidate for the specified vote rank}
#'   }
"cambridge_clean"

#' Raw ballot data from a Minneapolis, MN RCV election
#'
#' The .rda version of a .csv from a 2013 Minneapolis, MN mayoral RCV
#' election. This data is included as an example of another ballot format
#' for use with ballot functions.
#'
#' @format A data frame with 80101 rows and 5 variables
#' @source \url{http://vote.minneapolismn.gov/www/groups/public/@clerk/documents/webcontent/2013-mayor-cvr.xlsx}
"minneapolis_mayor_2013"

#' Labelled ballot lookup data from a San Francisco RCV election.
#'
#' The .rda version of sf_bos_lookup.rda after the `label()`
#' function has been applied to it.
#'
#' @format A data frame with 644 rows and 7 variables
"sf_lookup_labelled"

#' Tabulated results from a San Francisco RCV election
#'
#' Results from the SF "Board of Supervisors, District 7" election
#' in 2016. Tabulated with rcv_tally.
#' rcv_tally(image = sf_bos_clean,
#'           rcvcontest = "Board of Supervisors, District 7")
#'
#' @format a data frame with 6 rows and 5 variables
"sf_7_results"


R/plot_functions.R
#' Creates a data frame for use with the alluvial package
#'
#' @param image A dataframe containing rcv election data
#' @param rcvcontest The election to calculate results for
#' @return A dataframe that counts how many ballots follow each unique "path"
#' of candidates through the election rounds
#' @examples
#' \dontrun{
#' make_alluvialdf(image = sf_bos_clean, rcvcontest = "Board of Supervisors, District 7")
#' }
#' @export

make_alluvialdf <- function(image, rcvcontest) {
  contest <- . <- candidate <- id <- frequency <- NULL
  # create df of all voting combinations in election
  init <- readable(image)
  if (!(missing(rcvcontest))) {
    init <- init %>% dplyr::filter(contest == rcvcontest)
  }
  init <- init %>%
    dplyr::filter(contest == rcvcontest) %>%
    dplyr::select(3:ncol(.)) %>%
    dplyr::count_(lapply(names(.), as.name), sort = T) %>%
    dplyr::ungroup() %>%
    dplyr::mutate(id = rownames(.)) %>%
    tidyr::gather(key = rank, value = candidate, !! 1:(ncol(.)-2)) %>%
    dplyr::mutate(rank = as.numeric(rank))

  # create a losers df from the results df

  elim <- data.frame(candidate = character())

  results <- rcv_tally(image, rcvcontest)

  for (j in 2:(ncol(results)-1)) {
    loser <- results %>%
      dplyr::select(candidate, j) %>%
      dplyr::filter(!(is.na(results[, j]))) %>%
      dplyr::filter(candidate != "NA") %>%
      dplyr::filter(.[, 2] == min(.[, 2])) %>%
      dplyr::select(candidate)

    elim <- rbind(elim, loser)
  }

  alluvialdf <- data.frame(matrix(nrow = 0, ncol = ncol(results)))
  col_names <- c()
  for (i in 1:(ncol(results) - 1)) {
    col_names <- append(col_names, paste0("round", i))
  }
  colnames(alluvialdf) <- append(col_names, "frequency")

  pathframe <- data.frame(matrix(ncol = ncol(results)))
  colnames(pathframe) <- colnames(alluvialdf)

  for (j in unique(init$id)) {
    path <- pathframe

    votepattern <- init %>% dplyr::filter(id == j)

    for (i in 1:(ncol(pathframe)-1)) {

      tempelim <- data.frame(candidate = character())
      if (i >= 2) tempelim <- elim[1:(i-1),]

      path[,i] <- votepattern %>%
        dplyr::filter(!(candidate %in% tempelim)) %>%
        dplyr::filter(rank == min(rank)) %>%
        dplyr::select(candidate)
    }

    path[,ncol(pathframe)] <- votepattern[1,"n"]

    alluvialdf <- rbind(alluvialdf, path)
  }

  # get final frequencies attempt

  names <- lapply(names(alluvialdf)[-ncol(alluvialdf)], as.name)
  alluvialdf <- alluvialdf %>%
    dplyr::group_by_(.dots = names) %>%
    dplyr::summarise(frequency = sum(frequency)) %>%
    dplyr::arrange(dplyr::desc(frequency))

  # alluvial function cannot have NAs in it, so replace with "NA"
  alluvialdf[is.na(alluvialdf)] <- "NA"

  return(alluvialdf)
}

#' Creates a data frame for use with the networkD3 package
#'
#' @param results A dataframe containing rcv election results in the format
#' produced by `rcv_tally()`
#' @return A list of 2 dataframes that can be used to construct sankey diagrams
#' via the networkD3 package's `sankeyNetwork()` function
#' @examples
#' make_d3list(results = sf_7_results)
#' @export
make_d3list <- function(results) {
  count <- . <- candidate <- cand_id <- NULL

  names <- results %>%
    tidyr::gather(key = round, value = count, !! 2:ncol(results)) %>%
    dplyr::filter(!is.na(count))

  names_id <- names %>%
    tibble::rownames_to_column("cand_id") %>%
    dplyr::mutate(cand_id = as.numeric(cand_id) - 1) %>%
    dplyr::select(cand_id, candidate, round) %>%
    tidyr::spread(key = round, value = cand_id, fill = NA) %>%
    dplyr::select(2:ncol(results)) %>%
    dplyr::arrange(.[, 1])

  names <- names %>% dplyr::select(candidate)

  for (j in 3:ncol(results)) {
    temp <- data.frame(results[, 1], results[, j] - results[, j-1])
    colnames(temp) <- c("candidate", paste0("transfer", j-2))
    if (j == 3) {transfers <- temp
    } else transfers <- dplyr::left_join(transfers, temp, by = "candidate")
  }

  transfers <- transfers %>% dplyr::select(-candidate)

  round_totals <- results %>% dplyr::select(-candidate)

  source <- data.frame(source = numeric())

  for (j in 1:(ncol(names_id) - 1)) {
    source_temp <- data.frame(source = numeric())
    source_temp1 <- names_id %>%
      dplyr::filter(!is.na(names_id[, j+1])) %>%
      dplyr::select(j)
    colnames(source_temp1) <- c("source")
    source_temp2 <- names_id %>%
      dplyr::filter(is.na(names_id[, j+1])) %>%
      dplyr::select(j)
    source_temp2 <- data.frame(source = rep(source_temp2[1,1], nrow(source_temp1)))
    source_temp <- rbind(source_temp1, source_temp2)
    source <- rbind(source, source_temp)
  }

  target <- data.frame(target = numeric())

  for (j in 2:ncol(names_id)) {
    target_temp <- names_id %>%
      dplyr::filter(!is.na(names_id[, j])) %>%
      dplyr::select(j)
    colnames(target_temp) <- c("target")
    target <- rbind(target, target_temp, target_temp)
  }

  value <- data.frame(value = numeric())

  for (j in 1:ncol(transfers)) {
    temp_round <- round_totals %>%
      dplyr::filter(!is.na(round_totals[, j+1])) %>%
      dplyr::select(j)
    colnames(temp_round) <- c("value")
    temp_transfers <- transfers %>%
      dplyr::filter(!is.na(transfers[, j])) %>%
      dplyr::select(j)
    colnames(temp_transfers) <- c("value")
    temp_value <- rbind(temp_round, temp_transfers)
    value <- rbind(value, temp_value)
  }

  values <- data.frame(source, target, value)

  d3list <- list(values = values, names = names)
  return(d3list)
}



R/other_functions.R
#' Alters ballot dataframe to have each row correspond to a single voter
#'
#' @param clean A tidy dataframe that comes from ballot_tidy()
#' @return A dataframe that can be easily read and understood by humans
#' @examples readable(sf_bos_clean)
#' @export

readable <- function(clean) {
  contest <- pref_voter_id <- vote_rank <- candidate <- NULL
  clean %>%
    dplyr::select(contest,
                  pref_voter_id,
                  vote_rank,
                  candidate) %>%
    tidyr::spread(key = vote_rank, value = candidate)
}


#' Summarizes which candidates were elected from the results tally
#'
#' @param results the results tabulated from the election in question
#' @param n the number of candidates being elected (defaults to 1)
#' @return a vector of the candidates successfully elected
#' @examples elected(sf_7_results, n = 1)
#' @export

elected <- function(results, n = 1) {
  results[1:n, 1]
}


#' Summarizes the approval rate of the eventual winner(s)
#'
#' Counts what proportion of voters approved of the election winners, defining
#' "approved" as a voter listing the candidate on their ballot
#'
#' @param results the tabulated election results
#' @param image the clean ballot image
#' @param rcvcontest (optional) The election to calculate results for. If the image
#' contains more than one unique contest, this must be supplied.
#' @param n the number of candidates being elected (defaults to 1)
#' @return a numerical vector of length 1
#' @examples approval(sf_7_results, sf_bos_clean,
#'                    rcvcontest = "Board of Supervisors, District 7", n = 1)
#' @export

approval <- function(results, image, rcvcontest, n = 1) {
  voters <- approved <- contest <- candidate <- pref_voter_id <- NULL
  if (length(unique(image$contest)) > 1) {
    image <- image %>% dplyr::filter(contest == rcvcontest)
  }
  voters <- readable(image)
  approved <- image %>%
    dplyr::filter(candidate %in% elected(results, n)) %>%
    dplyr::count(pref_voter_id)
  nrow(approved) / nrow(voters)
}


R/cleaning_functions.R
#1234567890123456789012345678901234567890123456789012345678901234567890123456789

#' Imports election data
#'
#' Takes data argument supplied, checks file type, and uses appropriate read-in
#' functions to import the data
#'
#' @param data The file, containing ballot or lookup data
#' @param header Whether the first row of the file is a header or not
#' @return A data frame
#' @examples
#' import_data("http://www.sfelections.org/results/20161108/data/20161206/20161206_masterlookup.txt",
#'  header = FALSE)
#' @export

import_data <- function(data, header) {
  if ("data.frame" %in% class(data)) {
    data
  }
  else if (tools::file_ext(data) == "txt") {
    readr::read_tsv(data, col_names = header)
  }
  else if (tools::file_ext(data) == "csv") {
    readr::read_csv(data, col_names = header)
  }
  else stop('incompatible data format')
}


#' Separates single column election data frames.
#'
#' Takes a data frame of a single column (i.e. sf_bos_ballot) and splits it
#' into usable named columns.
#'
#' @param data A data frame with a single column
#' @param image Whether the data is a "ballot" or "lookup" image
#' @param format A character string detailing the format. Current
#' supported formats are "WinEDS" and "ChoicePlus" (forthcoming), based on
#' common types of software used. Contact creators with suggestions for
#' more formats.
#' @return A data frame with multiple columns
#' @examples
#' label(data = sf_bos_ballot, image = "ballot", format = "WinEDS")
#' @export

label <- function(data, image, format) {
  X1 <- tally_type_id <- vote_rank <- V1 <- a <- b <- V4 <- V3 <- NULL
  `1` <- candidate_id <- pref_voter_id <- record_type <- NULL
  description <- V2 <- NULL
  if (image == "ballot" & format == "WinEDS") {
    data %>%
      tidyr::separate(X1, into = c("contest_id",
                                   "pref_voter_id",
                                   "serial_number",
                                   "tally_type_id",
                                   "precinct_id",
                                   "vote_rank",
                                   "candidate_id",
                                   "over_vote",
                                   "under_vote"),
                      sep = c(7,16,23,26,33,36,43,44)) %>%
      dplyr::mutate(tally_type_id = as.integer(tally_type_id),
                    vote_rank = as.integer(vote_rank))
  }

  else if(image == "ballot" & format == "ChoicePlus") {
    x <- data %>%
      tidyr::separate(V1, into = c("a","ward","precinct","b"),
                      sep = c(2,4,6),
                      remove = T) %>%
      dplyr::select(-a, -b) %>%
      tidyr::separate(V4, into = c("a", "1"),
                      sep = 3,
                      remove = T) %>%
      tidyr::unite(col = "contest",
                   V3, a,
                   sep = ", ") %>%
      dplyr::mutate(`1` = replace(`1`, which(`1` == ""), NA))

    colnames(x) <- c("ward","precinct", "style", "contest",
                     as.character(c(1:(ncol(x)-4))))
    tall <- x %>%
      tibble::rownames_to_column("pref_voter_id") %>%
      tidyr::gather(key = vote_rank,
                    value = candidate_id,
                    !! 6:(ncol(x)+1),
                    na.rm = T) %>%
      dplyr::mutate(candidate_id = stringr::str_replace_all(candidate_id,
                                                            "\\[[0-9]{1,2}\\]",
                                                            ""),
                    pref_voter_id = as.integer(pref_voter_id),
                    vote_rank = as.integer(vote_rank)) %>%
      dplyr::arrange(pref_voter_id, vote_rank)
    return(tall)
  }

  else if (image == "lookup" & format == "WinEDS") {
    data %>%
      tidyr::separate(X1, into = c("record_type",
                                   "id",
                                   "description",
                                   "list_order",
                                   "condidates_contest_id",
                                   "is_writein",
                                   "is_provisional"),
                      sep = c(10,17,67,74,81,82)) %>%
      dplyr::mutate(record_type = trimws(record_type),
                    description = trimws(description))
  }

  else if (image == "lookup" & format == "ChoicePlus") {
    data %>%
      dplyr::filter(V1 == "20") %>%
      dplyr::select(V2,V3) %>%
      dplyr::rename(id = V2,
                    candidate = V3)
  }

  else stop('incompatible ballot format')
}


#' Replaces number string codes in ballot with character strings from lookup
#'
#' Matches codes in the `contest_id`, `tally_type_id`, `precinct_id`, and
#' `candidate_id` columns in the labelled ballot with codes from the
#' `id` column in the labelled lookup, then replaces these codes with
#' character values from the `description` column in the lookup.
#'
#' @param ballot The labelled ballot data
#' @param lookup The labelled lookup data
#' @param format A character string detailing the format. Current
#' supported formats are "WinEDS" and "ChoicePlus" (forthcoming), based on
#' common types of software used. Contact creators with suggestions for
#' more formats.
#' @return The ballot data, but now "readable" so votes can be understood
#' @examples
#' \dontrun{
#' characterize(ballot = sf_ballot_labelled, lookup = sf_lookup_labelled,
#' format = "WinEDS")
#' }
#' @export

characterize <- function(ballot, lookup, format) {
  record_type <- tally <- description <- contest <- pref_voter_id <- NULL
  serial_number <- id <- precinct <- vote_rank <- candidate <- NULL
  over_vote <- under_vote <- candidate_id <- NULL
  if (format == "WinEDS") {
  candidates <- lookup %>%
    dplyr::filter(record_type == "Candidate") %>%
    dplyr::select(id, description) %>%
    dplyr::rename(candidate = description)
  contests <- lookup %>%
    dplyr::filter(record_type == "Contest") %>%
    dplyr::select(id, description) %>%
    dplyr::rename(contest = description)
  precincts <- lookup %>%
    dplyr::filter(record_type == "Precinct") %>%
    dplyr::select(id, description) %>%
    dplyr::rename(precinct = description)
  tallies <- lookup %>%
    dplyr::filter(record_type == "Tally Type") %>%
    dplyr::select(id, description) %>%
    dplyr::mutate(id = as.integer(id)) %>%
    dplyr::rename(tally = description)

  dplyr::left_join(ballot, candidates, by = c("candidate_id" = "id")) %>%
    dplyr::left_join(contests, by = c("contest_id" = "id")) %>%
    dplyr::left_join(precincts, by = c("precinct_id" = "id")) %>%
    dplyr::left_join(tallies, by = c("tally_type_id" = "id")) %>%
    dplyr::select(contest,
           pref_voter_id,
           serial_number,
           tally,
           precinct,
           vote_rank,
           candidate,
           over_vote,
           under_vote)
  }

  else if (format == "ChoicePlus") {
    dplyr::left_join(ballot, lookup, by = c("candidate_id" = "id")) %>%
      dplyr::mutate(candidate = ifelse(is.na(candidate), candidate_id, candidate)) %>%
      dplyr::select(-candidate_id)
  }
}

#' Master one-step cleaning function
#'
#' Wraps `import_data`, `label`, and `characterize` to clean the ballot
#' image in one step.
#'
#' @param ballot The raw ballot image
#' @param b_header Whether the ballot image has a header line or not
#' @param lookup The raw lookup image
#' @param l_header Whether the lookup image has a header line or not
#' @param format A character string detailing the format. Current
#' supported formats are "WinEDS" and "ChoicePlus" (forthcoming), based on
#' common types of software used. Contact creators with suggestions for
#' more formats.
#' @return The ballot data, but now "readable" so votes can be understood
#' @examples clean_ballot(ballot = sf_bos_ballot, b_header = TRUE,
#' lookup = sf_bos_lookup, l_header = TRUE, format = "WinEDS")
#' @export

clean_ballot <- function(ballot, b_header, lookup, l_header, format) {
  a <- import_data(data = ballot, header = b_header) %>%
    label(image = "ballot", format = format)
  b <- import_data(data = lookup, header = l_header) %>%
    label(image = "lookup", format = format)
  characterize(ballot = a, lookup = b, format = format)
}

#' Function for cleaning Minneapolis RCV data
#'
#' The Minneapolis data comes in a different form than SF or Cambridge data.
#' This function optimizes the process for formatting this data.
#' @param data The raw RCV data
#' @return The data formatted for use with rcv_tally
#' @examples clean_mn(minneapolis_mayor_2013)
#' @export

clean_mn <- function(data) {
  vote_rank <- candidate <- `1` <- `2` <- `3` <- pref_voter_id <- NULL
  colnames(data) <- c("precinct", "1", "2", "3", "count")
  a <- data %>%
    dplyr::select(1:4) %>%
    tibble::rownames_to_column("pref_voter_id") %>%
    tidyr::gather(key = vote_rank, value = candidate, `1`, `2`, `3`) %>%
    dplyr::mutate(candidate = ifelse(candidate %in% c("undervote", "overvote"),
                                     NA, candidate)) %>%
    dplyr::mutate(pref_voter_id = as.numeric(pref_voter_id),
                  vote_rank = as.numeric(vote_rank)) %>%
    dplyr::arrange(pref_voter_id)

}
